File: backend/app/api/drift.py
"""Project-scoped drift API (E4).

Two surfaces:
  * Manual trap refresh — ``POST /api/projects/{id}/drift/refresh-traps``
    spins fresh hallucination traps via the trap-refresh service. Honors
    ``?count=`` override + ``?simulate=true`` for dev where no LLM
    credentials are wired.
  * Queue triage — list pending rows, accept (→ append to gold_test),
    reject (→ mark with a note).

Distinct from the per-deployment ``/deployments/{id}/drift/check``
surface, which evaluates a live endpoint's pass-rate; this one is
project-scoped and writes into the review queue rather than the
DeploymentDriftCheck table.
"""

from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel, Field
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.models.gold_drift_review_queue import GoldDriftQueueStatus
from app.models.project import Project


router = APIRouter(
    prefix="/projects/{project_id}/drift",
    tags=["Drift"],
)


class RefreshTrapsResponse(BaseModel):
    project_id: int
    generated: int
    clusters_targeted: list[str]
    simulated: bool
    row_ids: list[int]


@router.post("/refresh-traps", status_code=201, response_model=RefreshTrapsResponse)
async def refresh_traps(
    project_id: int,
    count: int | None = Query(None, ge=1, le=20),
    simulate: bool = Query(False),
    db: AsyncSession = Depends(get_db),
):
    """Generate fresh hallucination traps targeting recent failure
    clusters; persists them to the drift review queue.

    Query params:
      * ``count`` — override the per-project default (1-20).
      * ``simulate`` — bypass the LLM and use deterministic placeholder
        traps. Useful for dev + tests where no API key is wired.

    Errors:
      * 404 — project not found.
      * 400 — project has no recipe selected.
    """
    from app.services.drift_trap_refresh_service import refresh_traps_for_project

    try:
        return await refresh_traps_for_project(
            db,
            project_id=project_id,
            count=count,
            simulate=simulate,
        )
    except ValueError as exc:
        code = str(exc)
        if code == "project_not_found":
            raise HTTPException(404, code)
        raise HTTPException(400, code)


@router.get("/review-queue")
async def list_drift_review_queue(
    project_id: int,
    status: GoldDriftQueueStatus | None = Query(GoldDriftQueueStatus.PENDING),
    limit: int = Query(50, ge=1, le=500),
    db: AsyncSession = Depends(get_db),
):
    """Newest-first list of drift-queue rows for the project. Filters
    by ``status`` (default ``PENDING``). Pass ``status=`` (empty) to
    see every row including triaged ones for audit."""
    project = await db.get(Project, project_id)
    if project is None:
        raise HTTPException(404, "project_not_found")

    from app.services.drift_trap_refresh_service import list_review_queue

    rows = await list_review_queue(
        db,
        project_id=project_id,
        status=status,
        limit=limit,
    )
    return {"project_id": project_id, "rows": rows}


class TriageRequest(BaseModel):
    accept: bool = Field(...)
    note: str | None = Field(default=None, max_length=2000)


@router.post("/review-queue/{row_id}/triage")
async def triage_drift_queue_row(
    project_id: int,
    row_id: int,
    payload: TriageRequest,
    db: AsyncSession = Depends(get_db),
):
    """Accept / reject one drift-queue row. Accepting appends the
    row to the project's gold_test JSONL; rejecting leaves the row
    in place with the user's reason captured.

    Errors:
      * 404 — row not found in this project.
      * 409 — row already triaged (no re-triage; rejected rows stay
        rejected, accepted rows stay accepted).
    """
    from app.services.drift_trap_refresh_service import triage_queue_row

    try:
        row = await triage_queue_row(
            db,
            project_id=project_id,
            row_id=row_id,
            accept=payload.accept,
            note=payload.note,
        )
    except ValueError as exc:
        code = str(exc)
        if code == "queue_row_not_found":
            raise HTTPException(404, code)
        if code == "queue_row_already_triaged":
            raise HTTPException(409, code)
        raise HTTPException(400, code)

    return {
        "id": row.id,
        "status": row.status.value,
        "triaged_at": row.triaged_at.isoformat() if row.triaged_at else None,
        "triage_note": row.triage_note,
    }

File: frontend/src/api/drift.ts
/**
 * Typed client for the drift API (E4 UI).
 *
 * Wraps the four endpoints the DriftReviewQueuePanel consults:
 *   - GET  /drift/settings    → opt-in flag + count
 *   - PUT  /drift/settings    → mutate opt-in + count
 *   - GET  /drift/review-queue → list queue rows
 *   - POST /drift/refresh-traps → manual generate trigger
 *   - POST /drift/review-queue/{row}/triage → accept / reject one row
 */

import api from './client';


export type DriftQueueStatus = 'pending' | 'accepted' | 'rejected';


export interface DriftQueueRow {
    id: number;
    project_id: number;
    source_drift_check_id: number | null;
    cluster_reason_code: string | null;
    cluster_signature: string | null;
    payload: Record<string, unknown>;
    status: DriftQueueStatus;
    source_confidence: string;
    triage_note: string | null;
    created_at: string;
    triaged_at: string | null;
}


export interface DriftQueueListResponse {
    project_id: number;
    rows: DriftQueueRow[];
}


export interface DriftSettings {
    project_id: number;
    enabled: boolean;
    count: number;
}


export interface RefreshTrapsResponse {
    project_id: number;
    generated: number;
    clusters_targeted: string[];
    simulated: boolean;
    row_ids: number[];
}


export interface DriftTriageResponse {
    id: number;
    status: DriftQueueStatus;
    triaged_at: string | null;
    triage_note: string | null;
}


export async function fetchDriftSettings(projectId: number): Promise<DriftSettings> {
    const resp = await api.get(`/projects/${projectId}/drift/settings`);
    return resp.data as DriftSettings;
}


export async function updateDriftSettings(
    projectId: number,
    patch: { enabled?: boolean; count?: number },
): Promise<DriftSettings> {
    const resp = await api.put(`/projects/${projectId}/drift/settings`, patch);
    return resp.data as DriftSettings;
}


export async function listDriftReviewQueue(
    projectId: number,
    options: { status?: DriftQueueStatus; limit?: number } = {},
): Promise<DriftQueueListResponse> {
    const params: Record<string, string | number> = {};
    if (options.status !== undefined) params.status = options.status;
    if (options.limit !== undefined) params.limit = options.limit;
    const resp = await api.get(`/projects/${projectId}/drift/review-queue`, { params });
    return resp.data as DriftQueueListResponse;
}


export async function refreshDriftTraps(
    projectId: number,
    options: { count?: number; simulate?: boolean } = {},
): Promise<RefreshTrapsResponse> {
    const params: Record<string, string | number | boolean> = {};
    if (options.count !== undefined) params.count = options.count;
    if (options.simulate !== undefined) params.simulate = options.simulate;
    const resp = await api.post(
        `/projects/${projectId}/drift/refresh-traps`,
        null,
        { params },
    );
    return resp.data as RefreshTrapsResponse;
}


export async function triageDriftRow(
    projectId: number,
    rowId: number,
    payload: { accept: boolean; note?: string },
): Promise<DriftTriageResponse> {
    const resp = await api.post(
        `/projects/${projectId}/drift/review-queue/${rowId}/triage`,
        payload,
    );
    return resp.data as DriftTriageResponse;
}
